### DISCUSSION TITLE
On asset metadata files and a better integrated asset library

### DISCUSSION THREAD

**Discussion by theraot** [godotengine/godot-proposals#7990] [General Discussions]
The idea I'm going to share here is not meant for Godot 4, but it is a topic that cannot wait until Godot 5 is near, because as soon as Godot 5 gets a prerelease there will be submissions for the asset library. It also should be considered before there is an official asset store.

Here the idea: Asset metadata files. Some information form the submission form could be moved to asset metadata files, and new metadata could be added.

Goals:

- Ease the submission process and review, by having software parse asset metadata files.
- Allow more metadata on the assets that allows better integration on the engine. *With the expected effect that discoverability improves and there is less friction to use them.*
- Allow a repository to have multiple assets (please). *After all, we already can choose to not include files from an asset when we add it to a project.*

The asset metadata file would, of course, be automatically discarded by Godot, so it is not added to the project (that goes also for project template asset).

---

Here are some ideas of possible integrations:

- If the asset library can tell Godot which custom nodes are in an asset from the asset library, without downloading it※, we could have tab in the Create New Node window that is populated with nodes from the asset library.
- If the asset library can tell Godot which custom materials are in an asset from the asset library, without downloading it※, we could have an option in the inspector on any material property that opens a window that shows materials from the asset library (perhaps with static previews generated server side).
- *Instantiate child scene could have a tab populated with models from the asset library, addendum: custom themes both for the editor and when picking a resource, and so on.*

To be clear I'm not saying that every time we want to add a node or whatever Godot connects to the Internet. Which is why I say to have a separate tab, and only when you click there, Godot connects with the asset library.

Of course not every asset will have things to offer for those cases. So the asset library as is currently integrated in Godot should not go away.

※: Or download to a temporary or cache folder.

---

Even if the suggested integrations are not desired by the community, I believe there would still be value in a asset metadata file.

---
**Addendum on plugin.cfg**: Plugins already have a .cfg file, which is kind of metadata for them. But: 1. It is not surfaced (it is easier to read an standard path inside of the repository than to list the contents of the addons folder). 2. Assets can be templates or addons, and not every addon is a plugin.

**Addendum on proposals**

- The proposal https://github.com/godotengine/godot-proposals/issues/8114 by Juan Linietsky covers similar ground. However it also includes the idea of an "Endorsed" category of assets.
- The proposal https://github.com/godotengine/godot-proposals/issues/6934 by cassandrus (I had forgotten about it) also has the idea of integrating assets into the Create New Node window, however I believe having it be a separate tab is better. Calinou pointed out in a comment there that the asset library lacks the information to do this (hence, we need better metadata).

**Addendum on making a package manager**

A new metadata file for addons (and I remind you that not all addons are plugins), could also be used to express dependencies.

**Comment by passivestar** [comment]
Very similar to what I just proposed on twitter, so just so that I don't copy the whole text I'll leave a link here:
https://x.com/passivestar_/status/1816104174162964715

One thing that I want to add is that my proposal is not just technical, I also want there to be an agreement (endorsed with some kind of a "best practices" manual section or in some other way) in the community on how to make reusable components that others can use easily that would:

1. Try to be small like godot's built-in nodes
2. Try to be self-contained and have no dependencies. Or if dependencies are unavoidable they should communicate that clearly through the UI (i.e configuration warnings)
3. Have predictable interfaces that would not only make it easy to use them in code, but also allow anyone to use them with no code by linking events in editor, kind of like hammer entities
4. Have a clear documentation that would allow anyone to get started in seconds (this could be facilitated in the UI by showing documentation with main relevant methods and signals when you hover over scripts or packed scenes)

It would also be nice if this approach worked with editor scripts, i.e godot could have a script runner (which is needed on its own btw) and you would be able to quickly run editor scripts from your asset library
